README.md
# Site do xRAM-Memory

Este é o site para o projeto xRAM-Memory, um mecanismo de busca e arquivo de notícias sobre assuntos determinados assuntos da política.

## Entidades de dados

As peças de conteúdo do site podem ser divididas em:
- Entidades de conteúdo:
  - News (Notícia): uma notícia arquivada.
  - Document (Documento): um documento, que pode ser uma imagem, uma captura de notícia em pdf ou um arquivo em formato diverso
  - Page (Página estática): uma página com conteúdo em HTML arbitrário
  - Album (Álbum de fotos): um conjunto de Documentos do tipo imagem com uma página associada
- Entidades de taxonomia (classificação do conteúdo):
  - Keyword (Palavra-chave): uma simples palavra usada para descrever um conteúdo, exemplos: prisão, polícia federal, operação etc.
  - Subject (Assunto): um tópico que pode ter um texto descritivo e estar em destaque no site, por exemplo: "A prisão do presidente Temer".


## Estrutura do projeto

O site foi feito em [Nuxt.js](https://nuxtjs.org/), um framework para criação de sites em Vue.js.
A seguinte estrutura de arquivos foi adotada na pasta `src`:

```
├── assets
│   └── styles - estilos em formato dos processadores Sass e Stylus
├── components - os componentes em Vue utilizados no site
│   ├── common - componentes utilizados em vários locais
│   ├── home - componentes utilizados na página inicial
│   ├── nav - componentes de navegação
│   ├── news - componentes para exibição de informações de notícias
│   ├── search - componentes usados na página de busca
│   └── viewers - componentes para visualização de arquivos
├── config - configuração estática, com algumas constantes utilizadas por todo site
├── data-sources - serviços para busca de dados via HTTP
│   └── api - serviço para comunicação com a API em Django
├── layouts - componentes de layout, encapsulam os outros, utilizados pelas páginas
├── middleware - funções executadas a cada transição de página do site
├── pages - componentes vue que encapsulam as páginas do site
│   ├── album
│   │   └── _album_id.vue - página de um Álbum de Fotos, uma coleção de imagens
│   ├── contact.vue - página de contato
│   ├── document
│   │   └── _document_id.vue - página para um Documento individual, aberto num visualizador
│   ├── images.vue - página com os Álbuns de Fotos
│   ├── index.vue - página inicial
│   ├── news
│   │   └── _slug.vue - página individual de uma Notícia
│   ├── search.vue - página de pesquisa
│   ├── _slug.vue - página para uma Página Estática
│   ├── subject
│   │   └── _slug.vue - página de um Assunto
│   ├── subjects - página com os Assuntos e as Palavras-chave
│   └── subjects.vue - página com Assuntos e Palavras-chave
├── plugins
│   ├── contrib - componentes/bibliotecas de terceiros
│   └── custom - componentes/bibliotecas próprios/customizados
├── static - arquivos estáticos (imagens, favicons e metadados do site)
└── store - uma store Vuex
```

A pasta `test` conterá os testes automatizados, ainda não implementados nesta etapa do projeto.

## Rodando o projeto

### Localmente

Por padrão, o site espera que os seguintes serviços estejam rodando nos endereços abaixo:

- Backend em Django (API): `http://localhost:8000`
- Serviço relay de mensagens de contato: `http://localhost:3001`

O arquivo `.env.dist` é um template para um arquivo `.env` que deverá ser copiado para a pasta `./src`. Este arquivo permite configurar o endereço dos serviços acima tanto em *client-side* ou *server-side*, além de outras configurações - o arquivo está documentado e possui valores padrão.

O arquivo `.env` não é necessário para iniciar o site a menos que você queira alterar os valores padrão.

Uma vez configurado, o site pode ser iniciado em desenvolvimento com os comandos abaixo. O site rodará por padrão em `http://localhost:3000`

```bash
yarn dev
```
ou
```bash
npm run dev
```

Às vezes é interessante iniciar o site em modo SPA (*server side rendering* desabilitado) para ter uma visibilidade melhor dos erros, além de depurar código no navegador com mais facilidade. Para isso, rode:

```bash
yarn dev:spa
```
ou
```bash
npm run dev:spa
```

### Via Docker

CHANGELOG.md
Changelog
=========


v0.8.0 (2021-09-12)
-------------------

Corrigido
~~~~~~~~~
- Allow prefixed tags with a 'v' [Felipe]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.8.0. [Felipe]
- Feat/relevance sorting (#127) [Felipe Lube de Bragança]

  * chore: add relevance filter option

  * feat: sort by relevance (the original sort order)

  * fix: orderBy should be considered immediately on the first listing


v0.7.3 (2021-09-12)
-------------------

Adicionado
~~~~~~~~~~
- Image lazy loading in Card's default image. [Felipe]
- Set node version via .nvmrc. [Felipe]

Corrigido
~~~~~~~~~
- Avoid broken image symbol. [Felipe]
- Rename file. [Felipe]

Outros
~~~~~~
- Update packages and remove unused ones. [Felipe]

  Squashed commit of the following:

  commit 46a9f08a0cc54de30283f0de7b05292c210c469a
- Install a plugin for the lazysizes library. [Felipe]
- Add lazysizes library. [Felipe]
- Add npmrc file. [Felipe]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.7.3. [Felipe]


v0.7.2 (2021-04-09)
-------------------

Não categorizado
~~~~~~~~~~~~~~~~
- V0.7.2. [Felipe Lube de Bragança]


v0.7.1 (2021-04-09)
-------------------

Não categorizado
~~~~~~~~~~~~~~~~
- V0.7.1. [Felipe Lube de Bragança]


v0.7.0 (2021-04-08)
-------------------

Adicionado
~~~~~~~~~~
- Create a plugin to load the fonts. [Felipe Lube de Bragança]

Outros
~~~~~~
- Load the fonts via the plugin; remove external links. [Felipe Lube de
  Bragança]
- Add fonts used by the project. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.7.0. [Felipe Lube de Bragança]


v0.6.11 (2021-04-02)
--------------------

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.11. [Felipe Lube de Bragança]
- Revert "fix: center loading spinner" [Felipe Lube de Bragança]

  This reverts commit b098a8e4239a3a002ec81ce00823c680fea233c8.


v0.6.10 (2021-04-02)
--------------------

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.10. [Felipe Lube de Bragança]


v0.6.9 (2021-04-02)
-------------------

Corrigido
~~~~~~~~~
- Center loading spinner. [Felipe Lube de Bragança]

Outros
~~~~~~
- Update packages. [Felipe Lube de Bragança]
- Update target browser list. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.9. [Felipe Lube de Bragança]
- Refactor: LocalSearchBase as renderless component. [Felipe Lube de
  Bragança]


v0.6.8 (2021-03-28)
-------------------

Corrigido
~~~~~~~~~
- DocumentPages wrong variable access. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.8. [Felipe Lube de Bragança]
- Refactor: protect whole documentPages with try/catch. [Felipe Lube de
  Bragança]


v0.6.7 (2021-03-28)
-------------------

Corrigido
~~~~~~~~~
- Search and filtering bug. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.7. [Felipe Lube de Bragança]


v0.6.6 (2021-03-28)
-------------------

Corrigido
~~~~~~~~~
- Remove Object.freeze. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.6. [Felipe Lube de Bragança]


v0.6.5 (2021-03-28)
-------------------

Corrigido
~~~~~~~~~
- Revert to direct access to searchState to aviod read only error.
  [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.5. [Felipe Lube de Bragança]


v0.6.4 (2021-03-27)
-------------------

Corrigido
~~~~~~~~~
- Update version. [Felipe Lube de Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.4. [Felipe Lube de Bragança]


v0.6.3 (2021-03-27)
-------------------

Corrigido
~~~~~~~~~
- HomTagCloud: guard against zero standard deviation. [Felipe Lube de
  Bragança]

Não categorizado
~~~~~~~~~~~~~~~~
- V0.6.3. [Felipe Lube de Bragança]


v0.6.2 (2021-03-27)
-------------------

Corrigido
~~~~~~~~~
- Update package version. [Felipe Lube de Bragança]
- Do not filter results if there is no valid filter selected (#121)
  [Felipe Lube de Bragança]
- Correção na geração do changelog. [Felipe Lube de Bragança]

Outros
~~~~~~
- Chore: update @nuxt/axios. [Felipe Lube de Bragança]
- Remove ava and boilerplate tests. [Felipe Lube de Bragança]
- Chore(deps): [security] bump node-fetch from 2.6.0 to 2.6.1 (#112)
  [dependabot-preview[bot]]

  Bumps [node-fetch](https://github.com/bitinn/node-fetch) from 2.6.0 to 2.6.1. **This update includes a security fix.**
  - [Release notes](https://github.com/bitinn/node-fetch/releases)
  - [Changelog](https://github.com/node-fetch/node-fetch/blob/master/docs/CHANGELOG.md)
  - [Commits](https://github.com/bitinn/node-fetch/compare/v2.6.0...v2.6.1)
- Chore(deps): [security] bump bl from 1.2.2 to 1.2.3 (#116)
  [dependabot-preview[bot]]

  Bumps [bl](https://github.com/rvagg/bl) from 1.2.2 to 1.2.3. **This update includes a security fix.**
  - [Release notes](https://github.com/rvagg/bl/releases)
  - [Commits](https://github.com/rvagg/bl/compare/v1.2.2...v1.2.3)
- Chore(deps): [security] bump ini from 1.3.5 to 1.3.8 (#119)
  [dependabot-preview[bot]]

  Bumps [ini](https://github.com/isaacs/ini) from 1.3.5 to 1.3.8. **This update includes a security fix.**
  - [Release notes](https://github.com/isaacs/ini/releases)
  - [Commits](https://github.com/isaacs/ini/compare/v1.3.5...v1.3.8)
- Chore(deps): [security] bump elliptic from 6.5.3 to 6.5.4 (#120)
  [dependabot-preview[bot]]

  Bumps [elliptic](https://github.com/indutny/elliptic) from 6.5.3 to 6.5.4. **This update includes a security fix.**
  - [Release notes](https://github.com/indutny/elliptic/releases)
  - [Commits](https://github.com/indutny/elliptic/compare/v6.5.3...v6.5.4)
- Chore(deps): [security] bump object-path from 0.11.4 to 0.11.5 (#114)
  [dependabot-preview[bot]]

  Bumps [object-path](https://github.com/mariocasciaro/object-path) from 0.11.4 to 0.11.5. **This update includes a security fix.**
  - [Release notes](https://github.com/mariocasciaro/object-path/releases)
  - [Commits](https://github.com/mariocasciaro/object-path/commits)
- Chore(deps): [security] bump dot-prop from 4.2.0 to 4.2.1 (#115)
  [dependabot-preview[bot]]


nuxt.config.js
import axios from 'axios'
const pkg = require('./package')
const WorkerPlugin = require('worker-plugin')

module.exports = {
  target: 'static',
  srcDir: 'src/',
  generate: {
    crawler: true,
    routes: async function () {
      const baseURL = process.env.API_URL || 'http://localhost:8000'
      const username = process.env.WEBSITE_USER
      const password = process.env.WEBSITE_PASSWORD

      const { data: { token = '' } } = await axios.post('/api/v1/auth_token', {
        username,
        password
      }, { baseURL })

      if (token) {
        console.log('Token de autenticação obtido com sucesso')
      }

      const axiosInstance = axios.create({
        baseURL,
        headers: {
          Authorization: `Token ${token}`
        }
      })

      return Promise.all([
        axiosInstance.get('api/v1/news')
      ])
        .then(responses => responses.map(response => response.data))
        .then(pages => pages.reduce((acc, val) => acc.concat(val), []))
        .then(pages => pages.map(page => `/news/${page.slug}`))
    }
  },
  /*
  ** Headers of the page
  */
  head: {
    title: pkg.name,
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'description', name: 'description', content: pkg.description },
      { name: 'msapplication-TileColor', content: '#008987' },
      { name: 'theme-color', content: '#ffffff' }
    ],
    link: [
      { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' },
      {
        rel: 'apple-touch-icon',
        sizes: '180x180',
        href: '/apple-touch-icon.png'
      },
      {
        rel: 'icon',
        type: 'image/png',
        sizes: '32x32',
        href: '/favicon-32x32.png'
      },
      {
        rel: 'icon',
        type: 'image/png',
        sizes: '16x16',
        href: '/favicon-16x16.png'
      },
      { rel: 'manifest', href: '/site.webmanifest' },
      { rel: 'mask-icon', href: '/safari-pinned-tab.svg', color: '#5bbad5' }
    ]
  },
  /*
  ** Customize the progress-bar color
  */
  loading: { color: '#AA0000' },

  /*
  ** Global CSS
  */

  /**
   * Plugins to load before mounting the App
   */
  buildModules: [
    // Simple usage
    '@nuxtjs/vuetify'
  ],
  plugins: [
    '@/plugins/custom/fontSource.js',
    { src: '@/plugins/contrib/resize-sensor.js', ssr: false },
    { src: '@/plugins/contrib/element-ui', ssr: false },
    { src: '@/plugins/contrib/drag-scroll', ssr: false },
    { src: '@/plugins/contrib/async-computed', ssr: false },
    { src: '@/plugins/contrib/infinite-scroll', ssr: false },
    '@/plugins/contrib/axios',
    '@/plugins/custom/dataSources',
    '@/plugins/custom/utils',
    '@/plugins/custom/essential-content.js',
    '@/plugins/contrib/lazy-sizes'
  ],

  /*
  ** Nuxt.js modules
  */
  modules: [
    // Doc: https://axios.nuxtjs.org/usage
    '@nuxtjs/dotenv', // TODO: remover
    '@nuxtjs/axios',
    '@nuxtjs/style-resources'
  ],
  styleResources: {
    stylus: './assets/styles/*.styl'
  },
  /*
  ** Axios module configuration
  */
  axios: {
    baseURL: 'http://localhost:8000',
    debug: !(process.env.NODE_ENV && process.env.NODE_ENV === 'production'),
    progress: false
  },
  publicRuntimeConfig: {
    version: '0.7.1',
    axios: {
      browserBaseURL: process.env.API_URL
    },
    lunrIndexURL: process.env.LUNR_INDEX_URL ||
      'http://localhost:8000/media/lunr_index/index.json',
    contactMessageRelayURL: process.env.CONTACT_MESSAGE_RELAY_URL ||
      'http://localhost:3001',
    recaptchaKey: process.env.RECAPTCHA_KEY,
    mediaURL: process.env.MEDIA_URL ||
      'http://localhost:8000',
    apiURL: process.env.API_URL ||
      'http://localhost:8000'
  },
  privateRuntimeConfig: {
    axios: {
      baseURL: process.env.SSR_API_URL || process.env.API_URL
    },
    lunrIndexURL: process.env.SSR_LUNR_INDEX_URL ||
      process.env.LUNR_INDEX_URL ||
      'http://localhost:8000/media/lunr_index/index.json',
    contactMessageRelayURL: process.env.SSR_CONTACT_MESSAGE_RELAY_URL ||
      process.env.CONTACT_MESSAGE_RELAY_URL ||
      'http://localhost:3001',
    recaptchaKey: process.env.RECAPTCHA_KEY,
    mediaURL: process.env.SSR_MEDIA_URL ||
      process.env.MEDIA_URL ||
      'http://localhost:8000',
    apiURL: process.env.SSR_API_URL ||
      process.env.API_URL ||
      'http://localhost:8000'
  },
  router: {
    middleware: ['collapseMenu', 'fetchMenuItems'],
    parseQuery: qs => {
      // FIXME: não importe 'query-string' aqui
      const queryString = require('query-string')
      const parsed = queryString.parse(qs)
      if (parsed) {
        return parsed
      }
      return {}
    }
  },
  /*
  ** Build configuration
  */
  build: {
    /*
    ** You can extend webpack config here
    */
    transpile: [
      '@fontsource/material-icons',
      '@fontsource/roboto',
      '@fontsource/material-icons-outlined',
      '@fontsource/vollkorn',
      '@fontsource/vollkorn-sc'
    ],
    terser: {
      extractComments: false // default was LICENSES
    },
    babel: {
      plugins: [
        [
          'component',
          {
            libraryName: 'element-ui',
            styleLibraryName: 'theme-chalk'
          }
        ]
      ]
    },
    extend (config, ctx) {
      config.output.globalObject = 'this'
      // suporte para webworkers
      if (ctx.isClient) {
        config.plugins.push(new WorkerPlugin())
      }

      // Run ESLint on save
      if (ctx.isDev && ctx.isClient) {
        config.devtool = 'source-map'
      }
      config.module.rules.forEach(rule => {
        if (String(rule.test) === String(/\.(png|jpe?g|gif|svg|webp)$/)) {
          // add a second loader when loading images
          rule.use.push({
            loader: 'image-webpack-loader',
            options: {
              svgo: {
                plugins: [
                  // use these settings for internet explorer for proper scalable SVGs
                  // https://css-tricks.com/scale-svg/
                  { removeViewBox: false },
                  { removeDimensions: true }
                ]
              }
            }
          })
        }
      })
    },
    telemetry: false
  }
}
